### PR TITLE
Undo and Redo implemented

### LINKED ISSUE #119: Create a feature to undo previous commands

### LINKED ISSUE #122: Create temporary memory for redo and undo

### LINKED ISSUE #123: Create redo command

### PR DESCRIPTION
**Pull request by sumhungyee**
closes #119, closes #122, closes #123

Create interface for a specialised stack which supports deletion of old/deep items within the stack, as well as temporary storage of popped items.

Create a specialised stack with a deque as the main stack (which stores previous operations for undo) and a temporary storage stack (which stores operations for redo). The deque automatically polls when there are more than 10 items, and therefore, users will only be able to undo for a maximum of 10 times, which saves memory.

Create a memory class (UndoMemory) which abstracts away the lower-level operations of the stack data structure.

Create generate-deep-copy methods for ReadOnlyDengueHotspotTracker to accommodate different iterations of the same file within the memory stack.

Create Undo and Redo commands, and error handling for when the redo/undo limit is reached.

Edit clear command to clear all at once.

Edit test cases to accommodate changes in the Model class, as well as others.

Allow integer arguments for undo/redo for number of steps to undo/redo

### CODE REVIEW

**Review by markusyeo — COMMENTED** [review]
Main comment is to fix consistency of newlines.

Also to consider, is there a meaningful message for failure redo/undo to display?

File: src/main/java/seedu/dengue/logic/commands/RedoCommand.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+public class RedoCommand extends Command {
+    public static final String COMMAND_WORD = "redo";
+    public static final String MESSAGE_SUCCESS = "Redo successful!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + " performs a redo operation, where an optional argument \n"
+            + "can be provided to indicate the number of operations. \n"
+            + "eg. redo 2";
+
+    private final int numberOfRedos;
+    public RedoCommand(int numberOfRedos) {
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/logic/commands/RedoCommand.java:20]: Can add a newline here

File: src/main/java/seedu/dengue/logic/commands/UndoCommand.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+public class UndoCommand extends Command {
+    public static final String COMMAND_WORD = "undo";
+    public static final String MESSAGE_SUCCESS = "Undo successful!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + " performs an undo operation, where an optional argument \n"
+            + "can be provided to indicate the number of operations. \n"
+            + "eg. undo 1";
+
+    private final int numberOfUndos;
+    public UndoCommand(int numberOfUndos) {
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/logic/commands/UndoCommand.java:20]: Here needs a newline too

Comment (markusyeo) [inline src/main/java/seedu/dengue/logic/commands/UndoCommand.java:20]: do you want to add some java docs here? same for redo

File: src/main/java/seedu/dengue/logic/commands/UndoCommand.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        this.numberOfUndos = numberOfUndos;
+    }
+    @Override
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/logic/commands/UndoCommand.java:23]: some newline here too please

File: src/main/java/seedu/dengue/logic/parser/RedoCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * A parser for the arguments of the redo command.
+ */
+public class RedoCommandParser implements Parser<RedoCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the RedoCommand
+     * and returns a RedoCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/logic/parser/RedoCommandParser.java:17]: extra newline here

File: src/main/java/seedu/dengue/model/ModelManager.java [hunk 5]
```diff
@@ -101,21 +134,38 @@ public boolean hasPerson(Person person) {
...
     }
 
     @Override
     public void addPerson(Person person) {
         dengueHotspotTracker.addPerson(person);
         updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
+        memory.saveNewLatest(this.dengueHotspotTracker);
+        updateFromMemoryStack();
+
+
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/model/ModelManager.java:148]: Consistent newline spacing please

File: src/main/java/seedu/dengue/model/ModelManager.java [hunk 6]
```diff
@@ -165,4 +215,8 @@ public boolean equals(Object obj) {
                 && filteredPersons.equals(other.filteredPersons)
                 && overview.equals(other.overview);
     }
+
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/model/ModelManager.java:218]: this too

File: src/main/java/seedu/dengue/storage/temporary/MemoryStack.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Checks if the memory stack is full.
+     * @return A boolean value.
+     */
+    public boolean isFull() {
+        return this.memory.size() > MAX_SIZE;
+    }
+
+
```
Comment (markusyeo) [inline src/main/java/seedu/dengue/storage/temporary/MemoryStack.java:88]: can tidy this too

**Review by markusyeo — CHANGES_REQUESTED** [review]
Not much issue with code, mainly code formatting. You can try just using Intellij auto formatting it should help to fix all of it. Also some functions no javadocs, i presume you'll be adding it later

**Comment by sumhungyee** [comment]
> Main comment is to fix consistency of newlines.
> 
> Also to consider, is there a meaningful message for failure redo/undo to display?

yes, cannot undo any further and latest iteration reached for redo, all in CommandException

File: src/main/java/seedu/dengue/logic/commands/UndoCommand.java (hunk 2 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/logic/commands/UndoCommand.java:20]: done

Comment (sumhungyee) [reply src/main/java/seedu/dengue/logic/commands/UndoCommand.java:20]: done

File: src/main/java/seedu/dengue/logic/commands/UndoCommand.java (hunk 3 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/logic/commands/UndoCommand.java:23]: done

File: src/main/java/seedu/dengue/logic/parser/RedoCommandParser.java (hunk 4 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/logic/parser/RedoCommandParser.java:17]: done

File: src/main/java/seedu/dengue/model/ModelManager.java (hunk 5 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/model/ModelManager.java:148]: done

File: src/main/java/seedu/dengue/model/ModelManager.java (hunk 6 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/model/ModelManager.java:218]: done

File: src/main/java/seedu/dengue/storage/temporary/MemoryStack.java (hunk 7 above)
Comment (sumhungyee) [reply src/main/java/seedu/dengue/storage/temporary/MemoryStack.java:88]: done

**Pull request merged**